Modification in BuilderAi to pickup all needed ressources

lexustec opened the pull request:
# Changes proposed in this pull request:
- added ability to track multi stack items. ( bigger than stacksize. )
- less walks for the Builder if its a large build
- less buildtime
for review ... prepared testmap with birch townhall and all needed items available at..

https://github.com/lexustec/minecolonies-testmap

just start the townhall build and observe the inventory of the BuilderGuy

Review please

universemodifier committed:
Change in BuilderAi to be a little bit more smart---still has a bug

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real values.

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real values.

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real values.

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Bugfix... endless loop at bigger buildings

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Bugfix... endless loop at bigger buildings

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=4634) <br/>All committers have signed the CLA.

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -245,8 +252,13 @@ private IAIState getNeededItem()
-
-            tryTransferFromPosToWorker(walkTo, needsCurrently);
+            // modified for debug and added the Stack itself as a workaround to get the "real Stacksize"
+            if(tryTransferFromPosToWorker(walkTo, needsCurrently, needsCurrentlyStack))
```
I guess you can remove this here now.

lexustec replied:
fixed

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1021,9 +1033,8 @@ else if (dumpOneMoreSlot())
-        this.itemsNiceToHave().forEach(this::isInHut);
-        // we dumped the inventory, reset actions done
+         this.itemsNiceToHave().forEach(this::isInHut);
+         // we dumped the inventory, reset actions done
```
guess you can roll back this change too

lexustec replied:
done

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1461,20 +1472,35 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
      */
-    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Predicate<ItemStack> predicate)
+
+    // modified to reflect the actual stack amount to transfer.  WIP
```
this comment can also go then I believe

lexustec replied:
done

lexustec replied:
done

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1461,20 +1472,35 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+        if(stackInfo.getAmount() > stackInfo.getItemStack().getMaxStackSize())
+        {
+            insertAmount = stackInfo.getItemStack().getMaxStackSize();
+        }else{
```
can you format this properly?
In the repo there is a formatting file for intellij you can setup to properly format this too.

lexustec replied:
formatting done

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java:
```diff
@@ -155,15 +199,37 @@ public IAIState pickUpMaterial()
+//            }
+//            Log.getLogger().info("But has "+ otheramount +" already in local inventory ");
+            //return getState();
+            return GATHERING_REQUIRED_MATERIALS;
```
I don't understand this change. If the worker has the item in the inventory, then it doesn't have to pick more up, ain't I right?

Raycoms replied:
(You can also remove all the commented out code here)

lexustec replied:
done

universemodifier committed:
Modified and moved AI code

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real needed values.

universemodifier committed:
Modified and moved AI code

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real needed values.

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -121,7 +123,13 @@
+     * What he currently might be needing but as full information.
+     * this is a wip workaround... to get Stack Count right
+     */
+    protected BuildingBuilderResource needsCurrentlyStack = null;
```
If we're using `BuildingBuilderResource` here (not the builder class) we should change the name of the class

lexustec replied:
removed it. was used before the latest commit.

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -121,7 +123,13 @@
      * What he currently might be needing.
      */
-    protected Predicate<ItemStack> needsCurrently = null;
+    protected Tuple<Predicate<ItemStack>, Integer> needsCurrently = null;
```
Is there a reason we need a tuple here?
From what i can tell the Integer is for how much of an item we need, except that an ItemStack can store the count of items, and no, it's not restricted to 64, can be any Int.

lexustec replied:
Yes the Tuple is needed because only the Predicate is inside. Predicate<ItemStack>. this doesnt contain the count. the Tuple <Predicate<ItemStack>, Integer> was suggested by Raycoms.

lexustec commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/enchanter/EntityAIWorkEnchanter.java:
```diff
@@ -136,7 +136,7 @@ private IAIState decide()
                 if (numberOfBooksInBuilding > 0)
                 {
-                    needsCurrently = IS_BOOK;
+                    needsCurrently = new com.minecolonies.api.util.Tuple<>(IS_BOOK,1);
```
needs to be like this because the minecraft Tuple is used also. can not include both.

Asherslab replied:
Then please switch out the other tuple for minecraft tuple

lexustec replied:
if i change 139 to new Tuple<> it errors out with can not infere arguments.. if i change the import also at the top Line 237 errors with incompatible values. the com.minecolonies.api.util. was putted by intellij automaticly

lexustec replied:
Modified to the minecraft version.

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -245,8 +253,13 @@ private IAIState getNeededItem()
-
-            tryTransferFromPosToWorker(walkTo, needsCurrently);
+            // modified for debug and added the Stack itself as a workaround to get the "real Stacksize"
+            if(tryTransferFromPosToWorker(walkTo, needsCurrently))
```
If this was done for debug, and it does nothing, then please remove it for live. 
No need to keep it

lexustec replied:
comment line needs to be changed

Asherslab replied:
This should still be removed (nothing inside the if/else cases)

lexustec replied:
already modified

universemodifier committed:
formatting change

fixed formatting

universemodifier committed:
formatting change

fixed formatting

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1475,55 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
-          predicate,
-          Constants.STACKSIZE,
-          worker.getInventoryCitizen());
+        int existingAmount = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(),predicate.getA());
```
put a `final` on here please

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1475,55 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+        }
+
+        do{
+            int transferamount = 0;
```
incorrect casing here

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1475,55 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+            {
+                return false; // couldn't transfer
+            }
+        }while(amount > 0);
```
formatting (as ray said we have a formatting file)

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java:
```diff
@@ -77,7 +80,7 @@
      * After how many actions should the builder dump his inventory.
      */
-    private static final int ACTIONS_UNTIL_DUMP = 1024;
+    private static final int ACTIONS_UNTIL_DUMP = 4096;
```
Why are we increasing this?

lexustec replied:
if the Builder collected all materials for build we have  about 20 stacks in the inventory. which means if he is going to place all these 1280 blocks he will run back before placing all and dump the inventory. he will dump it anyway between the buildstages. also when he picks more material. I increased it to avoid this behaviour of unnescessary walks.

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/enchanter/EntityAIWorkEnchanter.java:
```diff
@@ -136,7 +136,7 @@ private IAIState decide()
                 if (numberOfBooksInBuilding > 0)
                 {
-                    needsCurrently = IS_BOOK;
+                    needsCurrently = new com.minecolonies.api.util.Tuple<>(IS_BOOK,1);
```
import at top of file please

lexustec replied:
can not import at top because conflict with minecraft.Tuple in Line 239. if i change 139 to new Tuple<> it errors out with can not infere arguments.. if i change the import also at the top Line 237 errors with incompatible values. the com.minecolonies.api.util. was putted by intellij automaticly

Asherslab commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/enchanter/EntityAIWorkEnchanter.java:
```diff
@@ -158,7 +158,7 @@ private IAIState decide()
             if (amountOfAncientTomes > 0)
             {
-                needsCurrently = IS_ANCIENT_TOME;
+                needsCurrently = new com.minecolonies.api.util.Tuple<>(IS_ANCIENT_TOME,1);
```
same as above

lexustec replied:
same as above

lexustec replied:
Tuple replaced now

Asherslab reviewed (changes requested):
Looks fine logic wise, but everything needs formatting (Yes including after newest commit)

universemodifier committed:
Fix and format

universemodifier committed:
Fix and format

universemodifier committed:
Format Fix

Fixed again. now should be good

universemodifier committed:
Format Fix

Fixed again. now should be good

universemodifier committed:
Tuple replace

Now it should be ready for review

universemodifier committed:
Tuple replace

Now it should be ready for review

universemodifier committed:
Revert "Warehouse Inventory Total page working"

This reverts commit e570e69500186e03ca671be786c93e668cf62028.

universemodifier committed:
Warehouse Inventory Total page working

The Warehouse got a new page which lists now all Warehouse connected Inventory. Icons still missing and window text is abstract for now. I didnt want to fuzz in the Language files for now. correct name needs to be found. Tech wise works nice. Ready for test

universemodifier committed:
Revert "Warehouse Inventory Total page working"

This reverts commit e570e69500186e03ca671be786c93e668cf62028.

universemodifier committed:
Warehouse Inventory Total page working

The Warehouse got a new page which lists now all Warehouse connected Inventory. Icons still missing and window text is abstract for now. I didnt want to fuzz in the Language files for now. correct name needs to be found. Tech wise works nice. Ready for test

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

universemodifier committed:
changed Function Name

changed the tryTransferFromPosToWorker to tryTransferFromPosToWorkerIfNeeded

universemodifier committed:
changed Function Name

changed the tryTransferFromPosToWorker to tryTransferFromPosToWorkerIfNeeded

universemodifier committed:
Building Hiring bug

Building hiring was able even if already worker was present.

universemodifier committed:
Building Hiring bug

Building hiring was able even if already worker was present.

lexustec reviewed (commented):
Formatting changes comitted

universemodifier committed:
Added function to average out itemstack pickups for large builds

universemodifier committed:
Added function to average out itemstack pickups for large builds

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

universemodifier committed:
hotfix stacksize

universemodifier committed:
hotfix stacksize

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

Raycoms commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/4634/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

LDTTeam-Buildsystem commented:
This PR has been automatically built and hosted by our build system, during this process there was a crash, error, or otherwise, please check on this at: https://teamcity.minecolonies.com/viewLog.html?buildId=37193&buildTypeId=LetSDevTogether_Minecolonies_PullRequests_2_DeployTestServer

marchermans commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1435,58 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+
+        int transferamount = 0;
+
+        do
```
Please adapt the format. While on top.

lexustec replied:
the do {} while is intended because it runs at least once.
a while {} do not.

lexustec replied:
I change it because at this point it is working both ways

marchermans commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java:
```diff
@@ -128,15 +132,55 @@ public IAIState getStateAfterPickUp()
-        final List<Predicate<ItemStack>> neededItemsList = new ArrayList<>();
-        for (final BuildingBuilderResource stack : building.getNeededResources().values())
+        final List<Tuple<Predicate<ItemStack>, Integer>> neededItemsList = new ArrayList<>();
+        boolean ONLY_64 = false;
```
This is not a constant. Adapt formatting.

lexustec replied:
changing

marchermans commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java:
```diff
@@ -128,15 +132,55 @@ public IAIState getStateAfterPickUp()
+        }
+        else
+        {
+            int stackCount = 0;
```
This need to be reworked. The logic here does not take stackability of items which are already contained in the inventory of the worker into account. Also formatting of while loops is of.

Raycoms replied:
Resolved

Raycoms committed:
Happiness (#4651)

Complete happiness rework:

Added four different types of Happiness Modifiers.

Static: Depending statically on a value of the citizen
Time Based: Change factor over time
Expiration based: Invoked for a short time (buff for three days on a colonist death for example)
Client Sided: With the factor only for the client side.

All the lists in the GUI are now automatically generated from the list of modifiers.

universemodifier committed:
Fixes

small changes,

universemodifier committed:
Fixes

small changes,

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15

marvin-bitterlich commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5356146&bid=17550900)

Raycoms committed:
Feature/code smell (#4659)

Add a bit more doc
Fix some code smell

Raycoms committed:
hotfix colony load issue

Raycoms committed:
Update building itemStorage (#4661)

* Update building itemStorage
* hotfix #4649

Raycoms committed:
hotfix #4662

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15
